File: epos_restaurant_2023/selling/report/sale_by_location/sale_by_location.py
# Copyright (c) 2024, Tes Pheakdey and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
	return get_columns(filters), get_data(filters)

def get_data(filters):
	filter  = "a.docstatus = 1"
	if filters.province:
		filter += " AND b.province IN ({0})".format(get_list(filters,"province"))
	if filters.district:
		filter += " AND b.district IN ({0})".format(get_list(filters,"district"))
	if filters.commune:
		filter += " AND b.commune IN ({0})".format(get_list(filters,"commune"))
	if filters.village:
		filter += " AND b.village IN ({0})".format(get_list(filters,"village"))
	sql = """
			SELECT
				coalesce(b.province,'') province,
				coalesce(b.district,'') district,
				coalesce(b.commune,'') commune,
				coalesce(b.village,'') village,
				SUM(c.quantity) quantity,
				SUM(c.sub_total) sub_total,
				SUM(c.total_discount) total_discount,
				SUM(c.total_tax) total_tax,
				SUM(c.total_revenue) total_revenue,
				SUM(c.cost*c.quantity) total_cost,
				SUM(c.total_revenue - (c.cost*c.quantity)) total_profit
			FROM `tabSale` a
			INNER JOIN `tabSale Product` c ON c.parent = a.name
			INNER JOIN `tabCustomer` b ON b.name = a.customer
			WHERE  {0}
			GROUP BY
				coalesce(b.province,''),
				coalesce(b.district,''),
				coalesce(b.commune,''),
				coalesce(b.village,'')
			Order By
				coalesce(b.province,''),
				coalesce(b.district,''),
				coalesce(b.commune,''),
				coalesce(b.village,'')""".format(filter)
	data = frappe.db.sql(sql,as_dict=1)
	return data

def get_columns(filters):
	columns = [
		{
			"label": _("Province"),
			"fieldname": "province",
			"fieldtype": "Data",
			"width": 100,
		},
		{
			"label": _("District"),
			"fieldname": "district",
			"fieldtype": "Data",
			"width": 100,
		},
		{
			"label": _("Commune"),
			"fieldname": "commune",
			"fieldtype": "Data",
			"width": 100,
		},
		{
			"label": _("Village"),
			"fieldname": "village",
			"fieldtype": "Data",
			"width": 100,
		},
		{
			"label": "Total Quantity",
			"fieldname": "quantity",
			"fieldtype": "Float",
			"width": 120,
		},
		{
			"label": "Sub Total",
			"fieldname": "sub_total",
			"fieldtype": "Currency",
			"width": 100,
		},
		{
			"label": "Total Discount",
			"fieldname": "total_discount",
			"fieldtype": "Currency",
			"width": 130,
		},
		{
			"label": "Total Tax",
			"fieldname": "total_tax",
			"fieldtype": "Currency",
			"width": 100,
		},
			{
			"label": "Total Revenue",
			"fieldname": "total_revenue",
			"fieldtype": "Currency",
			"width": 150,
		},
			{
			"label": "Total Cost",
			"fieldname": "total_cost",
			"fieldtype": "Currency",
			"width": 100,
		},
			{
			"label": "Total Profit",
			"fieldname": "total_profit",
			"fieldtype": "Currency",
			"width": 100,
		},
	]
	
	return columns

def get_list(filters,name):
	data = ','.join("'{0}'".format(x.replace("'", "''")) for x in filters.get(name))
	return data

File: epos_restaurant_2023/selling/report/sale_by_location/sale_by_location.js
// Copyright (c) 2024, Tes Pheakdey and contributors
// For license information, please see license.txt

frappe.query_reports["Sale By Location"] = {
	"filters": [
		{
			"fieldname":"branch",
			"label": __("Busuiness Branch"),
			"fieldtype": "Link",
			"options": "Business Branch",
			"reqd": 1,
			on_change: function (query_report) {},
		},
		{
			"fieldname":"start_date",
			"label": __("Start Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			"reqd": 1,
			"width": "60px",
			on_change: function (query_report) {},
		},
		{
			"fieldname":"end_date",
			"label": __("End Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1,
			"width": "60px",
			on_change: function (query_report) {},
		},
		{
			"fieldname":"province",
			"label": __("Province"),
			"fieldtype": "MultiSelectList",
			"options": "Province",
			get_data: function(txt) {
				return frappe.db.get_link_options('Province', txt, {});
			},
			on_change: function (query_report) {},
		},
		{
			"fieldname":"district",
			"label": __("District"),
			"fieldtype": "MultiSelectList",
			"options": "District",
			get_data: function(txt) {
				return frappe.db.get_link_options('District', txt, {});
			},
			on_change: function (query_report) {},
		},
		{
			"fieldname":"commune",
			"label": __("Commune"),
			"fieldtype": "MultiSelectList",
			"options": "Commune",
			get_data: function(txt) {
				return frappe.db.get_link_options('Commune', txt, {});
			},
			on_change: function (query_report) {},
		},
		{
			"fieldname":"village",
			"label": __("Village"),
			"fieldtype": "MultiSelectList",
			"options": "Village",
			get_data: function(txt) {
				return frappe.db.get_link_options('Village', txt, {});
			},
			on_change: function (query_report) {},
		},
	],
	onload: function(report) {
		report.page.add_inner_button("Preview Report", function () {
			frappe.query_report.refresh();
		});
		
	},
};
